Majority of refactoring to integrate ct-core-api.

jcrafford opened the pull request:
Still need to implement todo_resources and test.

jcrafford committed:
Majority of refactoring to integrate ct-core-api.
Still need to implement todo_resources and test.

jcrafford committed:
app --> todo_app
consider renaming back to app but in the meanwhile it makes the diffs easier.

bmorgan21 commented on flask/requirements.txt:
```diff
@@ -1,12 +1,4 @@
-Flask==0.11.1
```
shouldn't we still be pinning these versions?

jcrafford replied:
Still pinned, the ones that were removed come from `ct-core-api`

bmorgan21 commented on flask/todo_app/README.rst:
```diff
@@ -0,0 +1,31 @@
+Quick Start
+===========
+
+Set environment vars for the Flask application::
```
given we are in a docker container, shouldn't this speak to that?

jcrafford replied:
These instructions are for a local development setup that just uses Python and Flask.
Might be helpful on it's own but not as relevant with the code mounted in a docker container. Open to adding more context or removing.

bmorgan21 commented on flask/todo_app/extensions/principal_ext.py:
```diff
@@ -0,0 +1,8 @@
+from flask_principal import Principal
```
I really don't like having the directory information duplicated in the filename.  There has to be a better way to solve the problem.

jcrafford replied:
We touched on this in chat. Let me know if you still want to see changes.

bmorgan21 commented on flask/todo_app/models/user_models.py:
```diff
@@ -1,10 +1,12 @@
 
 
-class User(db.Model):
+class User(db.Model, UserMixin):
```
awesome!  cleans up the model!

bmorgan21 commented on flask/todo_app/models/user_models.py:
```diff
@@ -47,7 +28,4 @@ class Role(db.Model):
 
     @classmethod
     def get_all_for_user_id(cls, user_id):
```
all these classmethods that are doing queries in models should be removed.  I can clean that up after we get it merged.

bmorgan21 commented on flask/todo_app/modules/todo/todo_views.py:
```diff
@@ -1,10 +1,13 @@
 
 
 @bp.route("/")
+@app_utils.template()
```
what does this do?  does it default to index.html for the template by default?

jcrafford replied:
It renders the template with the name of the function + ".html" by default.
If the function returns a dict, it'll be used as the template vars.
Just a convenience decorator.

bmorgan21 commented on flask/todo_app/services/email_service.py:
```diff
@@ -1,16 +1,12 @@
-                  sender=DEFAULT_SENDER,
-                  recipients=[user.email])
-
+    msg = Message("Reset Your Password", sender=DEFAULT_SENDER, recipients=[user.email])
```
what is your rule for single line vs multiple line?  sometimes I agree w/ you and others I am not sure.

bmorgan21 commented on flask/todo_app/services/todo_service.py:
```diff
@@ -0,0 +1,13 @@
+__all__ = ['TodoService']
+
+
+class TodoManager(SqlAlchemyModelManager):
```
I think managers should be separate.  It is possible that multiple services will leverage the same manager.  i.e. to handle service composition or just things that require transactional integrity.

bmorgan21 reviewed (commented):
This stuff looks great!  a couple of comments for us to talk through.  Can I actually run it or are there other things to still handle?  i.e. are the ct-\* python libs public?

jcrafford committed:
todo_app --> app
Standardize on "app" top-level folder. If there were multiple apps then qualifying them would be appropriate.

jcrafford committed:
Checkpoint the todo resource conversion. Mostly done. Need to move onto testing.

jcrafford committed:
Add todo api error handlers for common exceptions.
minor naming fix.

bmorgan21 committed:
use ModalManager lib, refactor managers into their own thing.  not ideal to tie one-to-one with services as multiple services could leverage a manager

jcrafford committed:
Adjust entrypoints for the Flask app.

bmorgan21 committed:
make flask build faster by trading off image size. get things running
